File: src/main/java/cc/suvankar/moneytrail/contact/ContactService.java
package cc.suvankar.moneytrail.contact;

import cc.suvankar.moneytrail.contact.dto.ContactRequest;
import cc.suvankar.moneytrail.contact.dto.ContactResponse;
import cc.suvankar.moneytrail.exception.InvalidCredentialsException;
import cc.suvankar.moneytrail.exception.ResourceNotFoundException;
import cc.suvankar.moneytrail.user.User;
import cc.suvankar.moneytrail.user.UserService;
import java.util.List;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Slf4j
public class ContactService {

  private final ContactRepository contactRepository;
  private final UserService userService;

  public ContactService(ContactRepository contactRepository, UserService userService) {
    this.contactRepository = contactRepository;
    this.userService = userService;
  }

  @Transactional(readOnly = true)
  public Contact getContact(@NonNull UUID userId, @NonNull UUID contactId) {
    return contactRepository
        .findByUserIdAndId(userId, contactId)
        .orElseThrow(ResourceNotFoundException::forContact);
  }

  @Transactional(readOnly = true)
  public ContactResponse getContactResponse(@NonNull UUID userId, @NonNull UUID contactId) {
    return ContactResponse.from(getContact(userId, contactId));
  }

  @Transactional(readOnly = true)
  public List<ContactResponse> getContactsByUserId(@NonNull UUID userId) {
    return contactRepository.findByUserId(userId).stream().map(Contact::toResponse).toList();
  }

  @Transactional
  public ContactResponse createContact(@NonNull UUID userId, @NonNull ContactRequest request) {
    User user;

    try {
      user = userService.getUserById(userId);
    } catch (ResourceNotFoundException e) {
      throw new InvalidCredentialsException("Invalid credentials.");
    }

    var contact = new Contact();
    contact.setUser(user);
    contact.setName(request.getName());
    contact.setEmail(request.getEmail());
    contact.setPhoneNo(request.getPhoneNo());
    contact.setNotes(request.getNotes());

    log.info("Saving contact for user {}", userId);
    contact = contactRepository.save(contact);
    log.info("Saved contact {} for user {}", contact.getId(), userId);

    return ContactResponse.from(contact);
  }

  @Transactional
  public ContactResponse updateContact(
      @NonNull UUID userId, @NonNull UUID contactId, @NonNull ContactRequest request) {
    var contact = getContact(userId, contactId);

    contact.setName(request.getName());
    contact.setEmail(request.getEmail());
    contact.setPhoneNo(request.getPhoneNo());
    contact.setNotes(request.getNotes());

    log.info("Updating contact {} for user {}", contactId, userId);
    contact = contactRepository.save(contact);
    log.info("Updated contact {} for user {}", contactId, userId);

    return ContactResponse.from(contact);
  }

  @Transactional
  public void deleteContact(@NonNull UUID userId, @NonNull UUID contactId) {
    var contact = getContact(userId, contactId);
    log.info("Deleting contact {} for user {}", contactId, userId);
    contactRepository.delete(contact);
    log.info("Deleted contact {} for user {}", contactId, userId);
  }
}

File: src/main/kotlin/cc/suvankar/moneytrail/contact/ContactService.kt
package cc.suvankar.moneytrail.contact

import cc.suvankar.moneytrail.contact.dto.ContactRequest
import cc.suvankar.moneytrail.contact.dto.ContactResponse
import cc.suvankar.moneytrail.exception.ResourceNotFoundException
import cc.suvankar.moneytrail.user.UserService
import java.util.UUID
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class ContactService(
  private val contactRepository: ContactRepository,
  private val userService: UserService,
) {
  private val log = LoggerFactory.getLogger(ContactService::class.java)

  fun getContact(userId: UUID, contactId: UUID): Contact {
    return contactRepository
      .findByUserIdAndId(userId = userId, id = contactId)
      .orElseThrow(ResourceNotFoundException::forContact)
  }

  fun getContactsByUserId(userId: UUID): List<ContactResponse> {
    return contactRepository.findByUserId(userId = userId).map(ContactResponse::from)
  }

  fun createContact(userId: UUID, request: ContactRequest): ContactResponse {
    val contact =
      Contact(
        user = userService.getUserReferenceById(userId),
        name = request.name,
        email = request.email,
        phoneNo = request.phoneNo,
        notes = request.notes,
      )
    contactRepository.save(contact)

    log.info("Saving contact for user {}", userId)

    return ContactResponse.from(contact)
  }

  fun updateContact(userId: UUID, contactId: UUID, request: ContactRequest): ContactResponse {
    val contact =
      contactRepository
        .findByUserIdAndId(userId = userId, id = contactId)
        .orElseThrow(ResourceNotFoundException::forContact)

    contact.name = request.name
    contact.email = request.email
    contact.phoneNo = request.phoneNo
    contact.notes = request.notes

    log.info("Updating contact {} for user {}", contactId, userId)

    return ContactResponse.from(contact)
  }

  fun deleteContact(userId: UUID, contactId: UUID) {
    val contact =
      contactRepository
        .findByUserIdAndId(userId = userId, id = contactId)
        .orElseThrow(ResourceNotFoundException::forContact)

    log.info("Deleting contact {} for user {}", contactId, userId)

    contactRepository.delete(contact)
  }
}
